Using BoltDB

alexsporn opened the pull request.

luca-moser committed:
use BoltDB for storages which uses the ObjectStorage

alexsporn committed:
Ignore local hive.go

alexsporn committed:
- Adapting interfaces to be able to switch between badger and bolt

alexsporn committed:
Skip garbage collection with bolt

alexsporn committed:
- Use DeletePrefix instead of iterating and then deleting

muXxer commented on go.mod:
```diff
@@ -4,6 +4,8 @@ go 1.14
 
 replace github.com/dgraph-io/badger/v2 v2.0.1 => github.com/muXxer/badger/v2 v2.0.3-hotfix
 
+replace github.com/iotaledger/hive.go => ./hive.go
```
Do not forget to remove that at the end

muXxer commented on pkg/database/database.go:
```diff
@@ -0,0 +1,87 @@
+func GetDatabaseSize() (keys int64, values int64) {
+
+	if useBolt {
+		return 0, 0
```
Maybe we could use fileinfo here?

muXxer commented on pkg/database/database.go:
```diff
@@ -0,0 +1,87 @@
+	defer runtime.GC()
+
+	if useBolt {
+		return errors.New("Not available ")
```
`errors.New("not available")`

muXxer commented on pkg/database/database.go:
```diff
@@ -0,0 +1,87 @@
+	Entry     = database.Entry
+)
+
+func DatabaseWithPrefix(prefix byte) Database {
```
Comment

muXxer commented on pkg/database/database.go:
```diff
@@ -0,0 +1,87 @@
+	return badgerdb.NewDatabaseWithPrefix([]byte{prefix}, getBadgerInstance())
+}
+
+func StorageWithPrefix(prefix byte) objectstorage.Storage {
```
Comment

muXxer commented on pkg/database/database.go:
```diff
@@ -0,0 +1,87 @@
+	return getBadgerInstance().Size()
+}
+
+func Cleanup(discardRatio ...float64) error {
```
Comment

muXxer commented on pkg/database/database.go:
```diff
@@ -0,0 +1,87 @@
+	return cleanupBadgerInstance(discardRatio...)
+}
+
+func Close() error {
```
Comment

muXxer commented on pkg/model/tangle/tangle.go:
```diff
@@ -5,8 +5,8 @@ import (
 
-func ConfigureDatabases(directory string, badgerOpts *profile.BadgerOpts) {
-	database.Settings(directory, badgerOpts)
+func ConfigureDatabases(directory string, badgerOpts *profile.BadgerOpts, useBolt bool) {
```
Comment

alexsporn committed:
- Properly handling ErrKeyNotFound in database pkg

alexsporn committed:
- Remove the custom prefixes from the ledger db when streaming the keys

alexsporn committed:
- Remove badger and all its options

- Replaced database package with new store package that utilized the hive.go kvstore

alexsporn committed:
- Load maximum of 100 bundles in the dashboard explorer to avoid killing the node with bundle replay attacks

alexsporn committed:
- Removed store pkg

- Creating different bolt db files to separate tangle, snapshot and spent addresses

alexsporn committed:
- Properly close the autopeering db when shutting down the plugin

alexsporn committed:
- Fixed the db size metrics in the dashboard

alexsporn committed:
- Tuning bolt db options to have better performance for our usecase

alexsporn committed:
- Renamed lots of functions that had badger in the name

muXxer commented on plugins/database/plugin.go:
```diff
@@ -29,20 +23,7 @@ var (
 func configure(plugin *node.Plugin) {
 	log = logger.NewLogger(plugin.Name)
 
-	viper.BindEnv("GOMAXPROCS")
```
Shouldn't we set the GOMAXPROCS anyway? Maybe this also helps with SSD's and Bolt?

muXxer commented on plugins/database/plugin.go:
```diff
@@ -51,55 +32,44 @@ func configure(plugin *node.Plugin) {
-	return cleanups, err
-}
+func RunGarbageCollection() {
+	if tangle.DatabaseSupportCleanup() {
```
`DatabaseSupportsCleanup`

muXxer commented on pkg/model/tangle/health_db.go:
```diff
@@ -3,76 +3,64 @@ package tangle
 )
 
 const (
 	DbVersion = 9
```
Change to version 1

muXxer commented on pkg/model/tangle/ledger_db.go:
```diff
@@ -180,16 +141,19 @@ func GetLedgerDiffForMilestoneWithoutLocking(index milestone.Index, abortSignal
 		select {
 		case <-abortSignal:
-			return ErrOperationAborted
+			return false
```
We have to pass the `ErrOperationAborted` to the calling function, otherwise every user abort during this function would result in a panic because the sum won't be 0.

muXxer commented on pkg/model/tangle/ledger_db.go:
```diff
@@ -319,130 +278,29 @@ func ApplyLedgerDiffWithoutLocking(diff map[trinary.Hash]int64, index milestone.
-
-	// Now batch delete all entries
-	if err := ledgerDatabase.Apply([]database.Entry{}, deletions); err != nil {
+	// Delete ledger all balances
```
// Delete all ledger balances

muXxer commented on pkg/model/tangle/ledger_db.go:
```diff
@@ -486,18 +340,17 @@ func GetLedgerStateForLSMIWithoutLocking(abortSignal <-chan struct{}) (map[trina
 		select {
 		case <-abortSignal:
-			return ErrOperationAborted
+			return false
```
Same here, I guess we have to pass `ErrOperationAborted` to the calling function.

muXxer commented on pkg/model/tangle/tangle.go:
```diff
@@ -1,27 +1,112 @@
+	return ErrNothingToCleanUp
+}
+
+// GetDatabaseSizes returns the size of the database keys and values.
```
Wrong comment

muXxer committed:
Add YRTSI and ORTSI to transaction metadata

alexsporn committed:
- Updated to latest hive.go

alexsporn committed:
- Added back GOMAXPROCS support

alexsporn committed:
- Removed NoFreelistSync from the bolt options, since this means opening the db will take a lot of time, since the whole db needs to be scanned to build the freelist

alexsporn committed:
- Properly handle the abortion of the ledger iteration

alexsporn committed:
- Review changes

muXxer commented on pkg/model/tangle/snapshot_db.go:
```diff
@@ -111,7 +111,10 @@ func StoreSnapshotBalancesInDatabase(balances map[trinary.Hash]uint64, index mil
 		key := trinary.MustTrytesToBytes(address)[:49]
 		if balance != 0 {
-			batch.Set(key, bytesFromBalance(balance))
+			err := batch.Set(key, bytesFromBalance(balance))
```
```go
			if err := batch.Set(key, bytesFromBalance(balance)); err != nil {
				return errors.Wrap(NewDatabaseError(err), "failed to set the balance")
			}```

alexsporn committed:
- Prepare for rc9

alexsporn committed:
- Fix for Max

alexsporn committed:
- Run packr with latest changes

alexsporn committed:
- Woof woof

muXxer committed:
Fix spammer settings

muXxer committed:
Fixed changelog
